src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp

```diff
@@ -1840,6 +1840,74 @@ class AutoDiffSort
   }
 };
 
+class AutoDiffBatchNormTrainingRev
+    : public ReverseAutoDiffOpInterface::ExternalModel<
+          AutoDiffBatchNormTrainingRev, BatchNormTrainingOp> {
+public:
+  LogicalResult createReverseModeAdjoint(Operation *orig, OpBuilder &builder,
+                                         MGradientUtilsReverse *gutils,
+                                         SmallVector<Value> caches) const {
+    auto op = cast<BatchNormTrainingOp>(orig);
+
+    if (!gutils->isConstantValue(op->getResult(0)) ||
+        !gutils->isConstantValue(op->getResult(1)) ||
+        !gutils->isConstantValue(op->getResult(2))) {
+      gutils->zeroDiffe(op->getResult(0), builder);
+
+      auto opOperand0 = gutils->popCache(caches[0], builder);
```

Comment:
this looks like it ought be doable in tablegen, was there a reason why not?

Reply:
the 3 gradients are from the same grad op. I couldn't figure out how to do it without inserting 3 separate ops

Reply:
ah I see, yeah that makes sense. 

we can extend this in the future but this is good now